diffeqpy/_julia_project.py

```diff
@@ -0,0 +1,38 @@
+import julia
+import logging
+
+from julia_project import JuliaProject
+
+import os
+diffeqpy_path = os.path.dirname(os.path.abspath(__file__))
+
+julia_project = JuliaProject(
+    name="diffeqpy",
+    package_path=diffeqpy_path,
+    preferred_julia_versions = ['1.7', '1.6', 'latest'],
+    env_prefix = 'DIFFEQPY_',
+    logging_level = logging.INFO, # or logging.WARN,
+    console_logging=False
+)
+
+julia_project.run()
```

Comment:
> makes it impossible to provide APIs that can/should be used before initialization
I don't see at the moment when this would happen. But, I suspect you are correct. So, you would have to do `import diffeqpy`, then `diffeqpy.setup()` or something like that I suppose.

Reply:
> `import diffeqpy`, then `diffeqpy.setup()`

Yes, that's the idea. It'd be better if

```pycon
>>> import diffeqpy
>>> diffeqpy.update(julia=True)  # hypothetical API to update DifferentialEquations and Julia
>>> from diffeqpy import de
```

works. If we initialize things in `import diffeqpy`, it'd be impossible to use updated `julia` and DifferentialEquations (though the latter is somewhat possible if we integrate Revise).

I also consider that it's a language-agnostic best practice to avoid "magic" initialization on module/library import.

Re https://github.com/SciML/diffeqpy/pull/100#issuecomment-1010578461

> It installs packages and builds a system image in both the source dir and the environment created for tox.

I don't think this is related to the package initialization, though. Do you generate a sysimage for each libpython? If so, it might be due to that `tox` (or rather `virtualenv`) creates a Python executable and libpython for each environment (IIRC). I remember `venv` to be more "light weight" than `virtualenv` since it doesn't copy executable and libpython (IIRC). So maybe using `tox-venv` could help.

Reply:
> I also consider that it's a language-agnostic best practice to avoid "magic" initialization

Yes, the  idea of `import package` reading a bunch of state from the filesystem plus implicit inputs (environment variables) and then changing a bunch of state seems like its abusing the import system. I can't see an actual serious problem, but it doesn't feel right. I really like the idea of the user not being required to remember to do anything to use the package. But, requiring `setup` or `install` the first time is perhaps not so bad.

>  it'd be impossible to use updated julia and DifferentialEquations

You could require that the user restart after upgrading. I think people are pretty used to software that tells them they have to do that.

> Do you generate a sysimage for each libpython?

If there is a conflicting `libpython` then `julia_project` gives the user the option of either rebuilding PyCall (and breaking whatever built it previously) or using a depot particular to the project and building a new PyCall (and all other packages) there. (I don't know of a more fine-grained option at the moment) In the tox test, so far, I only choose the latter. But, I think it may not be that. I did not spend much time debugging tox so far.

Reply:
> But, requiring `setup` or `install` the first time is perhaps not so bad.

I'm OK with `from diffeqpy import de` working out-of-the-box by doing some magic by default. I'm just suggesting to provide an option to do some operations before booting up PyJulia and DifferentialEquations. In fact, this is exactly why diffeqpy does the initialization in `diffeqpy/de.py` and not in `diffeqpy/__init__.py`.

> You could require that the user restart after upgrading. I think people are pretty used to software that tells them they have to do that.

I don't think we need to copy badly designed software when there is a simple and better solution that is already implemented.